File: leetcode/editor/en/[1326]Minimum Number of Taps to Open to Water a Garden.java

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public int minTaps(int n, int[] ranges) {
        int[] interval = new int[ranges.length];
        for (int i = 0; i < ranges.length; i++) {
            int left = Math.max(0, i - ranges[i]);
            int right = i + ranges[i];
            interval[left] = Math.max(interval[left], right);
        }

        int currEnd = 0;
        int nextEnd = 0;
        int res = 0;
        for (int i = 0; i <= n; i++) {
            if (i > nextEnd) {
                return -1;
            }
            if (i > currEnd) {
                res++;
                currEnd = nextEnd;
            }
            nextEnd = Math.max(interval[i], nextEnd);
        }
        return res;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[1326]Minimum Number of Taps to Open to Water a Garden.js
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * @param {number} n
 * @param {number[]} ranges
 * @return {number}
 */
var minTaps = function (n, ranges) {
    let interval = new Array(ranges.length).fill(0);
    for (let i = 0; i < ranges.length; i++) {
        let left = Math.max(0, i - ranges[i]);
        let right = i + ranges[i];
        interval[left] = Math.max(interval[left], right);
    }

    let currEnd = 0;
    let nextEnd = 0;
    let res = 0;
    for (let i = 0; i <= n; i++) {
        if (i > nextEnd) {
            return -1;
        }
        if (i > currEnd) {
            res++;
            currEnd = nextEnd;
        }
        nextEnd = Math.max(nextEnd, interval[i]);
    }
    return res;
};
//leetcode submit region end(Prohibit modification and deletion)
